### ISSUE TITLE
Cannot change location of schema.prisma at runtime

### ISSUE THREAD

**Issue by royra** [prisma/prisma#2746]
Running on node with webpack.

Can specify the location of the query engine binary using `PRISMA_QUERY_ENGINE_BINARY`.

I didn't find a way to do the same for the `schema.prisma` file.

With webpack, code from `node_modules` is bundled into a single file. Runtime can't do `fs.readFile` from a `node_modules` dir and assume it's there.

This is the debug log:

```
 prisma-client Client Version 2.0.0-beta.7 +0ms
  prisma-client Engine Version 5d39801acf2e3475bd9dab029a63634358b07bf1 +2ms
2020-06-14T15:43:52.828Z info src/index starting wslc server
  prisma-client {
  prisma-client   engineConfig: {
  prisma-client     cwd: '../../node_modules/.prisma/client',
  prisma-client     debug: false,
  prisma-client     datamodelPath: '../../node_modules/.prisma/client/schema.prisma',
  prisma-client     prismaPath: undefined,
  prisma-client     generator: {
  prisma-client       name: 'client',
  prisma-client       provider: 'prisma-client-js',
  prisma-client       output: '/Users/roy/salto_private/node_modules/@prisma/client',
  prisma-client       binaryTargets: [Array],
  prisma-client       config: {}
  prisma-client     },
  prisma-client     showColors: false,
  prisma-client     logLevel: 'info',
  prisma-client     logQueries: false,
  prisma-client     flags: [],
  prisma-client     clientVersion: '2.0.0-beta.7'
  prisma-client   }
  prisma-client } +0ms
  prisma-client Prisma Client call: +59ms
  prisma-client prisma.raw(select 1) +1ms
  prisma-client Generated request: +15ms
  prisma-client mutation {
  prisma-client   executeRaw(query: "select 1")
  prisma-client }
  prisma-client  +0ms
  engine { cwd: '/project' } +0ms
  plusX Have to call plusX on /project/query-engine-linux-musl +0ms
  engine { flags: [ '--enable-raw-queries' ] } +98ms
  engine stderr thread 'main' panicked at 'Could not open datamodel file "../../node_modules/.prisma/client/schema.prisma": Os { code: 2, kind: NotFound, message: "No such file or directory" }', query-engine/query-engine/src/opt.rs:161:17 +88ms
  engine stderr stack backtrace: +0ms
  engine stderr    0: backtrace::backtrace::libunwind::trace +28ms
  engine stderr              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86 +0ms
  engine stderr    1: backtrace::backtrace::trace_unsynchronized +0ms
  engine stderr              at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66 +0ms
  engine stderr    2: std::sys_common::backtrace::_print_fmt +1ms
  engine stderr              at src/libstd/sys_common/backtrace.rs:78 +0ms
  engine stderr    3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt +1ms
  engine stderr              at src/libstd/sys_common/backtrace.rs:59 +0ms
  engine stderr    4: core::fmt::write +17ms
  engine stderr              at src/libcore/fmt/mod.rs:1063 +1ms
  engine stderr    5: std::io::Write::write_fmt +0ms
  engine stderr              at src/libstd/io/mod.rs:1426 +1ms
  engine stderr    6: std::sys_common::backtrace::_print +1ms
  engine stderr              at src/libstd/sys_common/backtrace.rs:62 +1ms
  engine stderr    7: std::sys_common::backtrace::print +1ms
  engine stderr              at src/libstd/sys_common/backtrace.rs:49 +1ms
  engine stderr    8: std::panicking::default_hook::{{closure}} +2ms
  engine stderr              at src/libstd/panicking.rs:204 +1ms
  engine stderr    9: std::panicking::default_hook +0ms
  engine stderr              at src/libstd/panicking.rs:224 +1ms
  engine stderr   10: std::panicking::rust_panic_with_hook +1ms
  engine stderr              at src/libstd/panicking.rs:470 +0ms
  engine stderr   11: rust_begin_unwind +1ms
  engine stderr              at src/libstd/panicking.rs:378 +0ms
  engine stderr   12: core::panicking::panic_fmt +0ms
  engine stderr              at src/libcore/panicking.rs:85 +1ms
  engine stderr   13: core::option::expect_none_failed +0ms
  engine stderr              at src/libcore/option.rs:1211 +0ms
  engine stderr   14: query_engine::main::{{closure}} +0ms
  engine stderr   15: tokio::runtime::enter::Enter::block_on +0ms
  engine stderr   16: tokio::runtime::thread_pool::ThreadPool::block_on +1ms
  engine stderr   17: tokio::runtime::context::enter +0ms
  engine stderr   18: query_engine::main +0ms
  engine stderr   19: std::rt::lang_start::{{closure}} +0ms
  engine stderr   20: std::rt::lang_start_internal::{{closure}} +1ms
  engine stderr              at src/libstd/rt.rs:52 +0ms
  engine stderr   21: std::panicking::try::do_call +0ms
  engine stderr              at src/libstd/panicking.rs:303 +0ms
  engine stderr   22: __rust_maybe_catch_panic +1ms
  engine stderr              at src/libpanic_unwind/lib.rs:86 +0ms
  engine stderr   23: std::panicking::try +0ms
  engine stderr              at src/libstd/panicking.rs:281 +0ms
  engine stderr   24: std::panic::catch_unwind +4ms
  engine stderr              at src/libstd/panic.rs:394 +1ms
  engine stderr   25: std::rt::lang_start_internal +0ms
  engine stderr              at src/libstd/rt.rs:51 +0ms
  engine stderr   26: main +3ms
  engine stderr note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace. +0ms
(node:11) UnhandledPromiseRejectionWarning: Error:
thread 'main' panicked at 'Could not open datamodel file "../../node_modules/.prisma/client/schema.prisma": Os { code: 2, kind: NotFound, message: "No such file or directory" }', query-engine/query-engine/src/opt.rs:161:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1063
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:470
  11: rust_begin_unwind
             at src/libstd/panicking.rs:378
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  13: core::option::expect_none_failed
             at src/libcore/option.rs:1211
  14: query_engine::main::{{closure}}
  15: tokio::runtime::enter::Enter::block_on
  16: tokio::runtime::thread_pool::ThreadPool::block_on
  17: tokio::runtime::context::enter
  18: query_engine::main
  19: std::rt::lang_start::{{closure}}
  20: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  21: std::panicking::try::do_call
             at src/libstd/panicking.rs:303
  22: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:86
  23: std::panicking::try
             at src/libstd/panicking.rs:281
  24: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  25: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  26: main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
    at R.request (/project/webpack:/Users/roy/salto_private/node_modules/@prisma/client/runtime/index.js:1:74271)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:11) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:11) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by janpio** [comment]
Do I understand correctly that you are using webpack in your app, and as a result our Rust Query Engine fails at reading `schema.prisma` at the usual location?

**Comment by royra** [comment]
> Do I understand correctly that you are using webpack in your app, and as a result our Rust Query Engine fails at reading `schema.prisma` at the usual location?

Exactly.

> (You might want to update from beta 7 to the current `2.0.0` release to make sure this was not fixed before that release. Be ware of the changes in beta 8 and 9 though when updating.)

2.0.0 logs are pretty much the same:

```
  prisma-client Client Version 2.0.0 +0ms
  prisma-client Engine Version de2bc1cbdb5561ad73d2f08463fa2eec48993f56 +1ms
2020-06-15T20:51:40.554Z debug process.cwd is "/project"
  prisma-client {
  prisma-client   engineConfig: {
  prisma-client     cwd: '../../node_modules/.prisma/client',
  prisma-client     debug: false,
  prisma-client     datamodelPath: '../../node_modules/.prisma/client/schema.prisma',
  prisma-client     prismaPath: undefined,
  prisma-client     generator: {
  prisma-client       name: 'client',
  prisma-client       provider: 'prisma-client-js',
  prisma-client       output: '/Users/roy/salto_private/node_modules/@prisma/client',
  prisma-client       binaryTargets: [Array],
  prisma-client       config: {}
  prisma-client     },
  prisma-client     showColors: false,
  prisma-client     logLevel: 'info',
  prisma-client     logQueries: false,
  prisma-client     flags: [],
  prisma-client     clientVersion: '2.0.0'
  prisma-client   }
  prisma-client } +0ms
  prisma-client Prisma Client call: +82ms
  prisma-client prisma.queryRaw(select 1) +2ms
  prisma-client Generated request: +24ms
  prisma-client mutation {
  prisma-client   queryRaw(query: "select 1")
  prisma-client }
  prisma-client  +0ms
  engine { cwd: '/project' } +0ms
  engine {
  engine   dotPrismaPath: '../../node_modules/.prisma/client/query-engine-linux-musl'
  engine } +35ms
  engine {
  engine   dirnamePath: '../../node_modules/@prisma/client/runtime/query-engine-linux-musl'
  engine } +0ms
  engine {
  engine   parentDirName: '../../node_modules/@prisma/client/query-engine-linux-musl'
  engine } +1ms
  plusX Have to call plusX on /project/query-engine-linux-musl +0ms
  engine { flags: [ '--enable-raw-queries' ] } +257ms
  engine stderr thread 'main' panicked at 'Could not open datamodel file "../../node_modules/.prisma/client/schema.prisma": Os { code: 2, kind: NotFound, message: "No such file or directory" }', query-engine/query-engine/src/opt.rs:161:17 +102ms
```

**Comment by royra** [comment]
> Any minimal reproduction for your webpack setup so our Support Engineers can start with something for their confirmation efforts?

[Here you go](https://github.com/salto-io/prisma-issue-2746).

**Comment by pantharshit00** [comment]
I will classify this as a bug in schema resolution as schema is correctly stored inside a `prisma` folder. 

We can also explore if we need to expose a env var for schema path. cc @timsuchanek

**Comment by heymartinadams** [comment]
Receiving same error ever since we had to create a custom hack ([manually integrate `react-refresh`](https://github.com/facebook/react/issues/16604#issuecomment-528663101)) in order to integrate Preact (not React) with the latest NextJS `9.5.4` and deploy it on Vercel (long story... 😅 )

Ever since then, somehow Prisma can’t find `./node_modules/@prisma/client/schema.prisma`, so we’ve added a copy into our `postinstall` script as a workaround. That seems to have worked.

```json
{
  "scripts": {
    "postinstall": "prisma generate && cp -r ./node_modules/.prisma/client/schema.prisma ./node_modules/@prisma/client"
  }
}
```

**Comment by ben-walker** [comment]
Also running into this with `@vercel/ncc` (which I believe uses webpack under the hood).

`schema.prisma` is sequestered away in `dist/client/`, and the build seems to want it to be in the root `dist/` directory.

**Comment by johannesschobel** [comment]
yeah, i am running into the same issue when using `nrwl/nx` as monorepository framework and `nestjs` for my API.

NX bundles the entire API into one large js file that is served. However, the `schema.prisma` file is not "copied" to the `dist/api` folder, hence, resulting in the previously mentioned error.

I am using the latest version of
- nrwl/nx
- nestjs
- prisma

Are there any suggestions? besides "manually" copying the files (or via npm `postXXX` scripts?)

All the best and thanks for your help,
Johannes

**Comment by alexmensak** [comment]
> yeah, i am running into the same issue when using `nrwl/nx` as monorepository framework and `nestjs` for my API.
> 
> NX bundles the entire API into one large js file that is served. However, the `schema.prisma` file is not "copied" to the `dist/api` folder, hence, resulting in the previously mentioned error.
> 
> I am using the latest version of
> 
> * nrwl/nx
> * nestjs
> * prisma
> 
> Are there any suggestions? besides "manually" copying the files (or via npm `postXXX` scripts?)
> 
> All the best and thanks for your help,
> Johannes

Hi,

did you managed to solve this please? I have ran to exactly same issue with `nx` and it's super frustrating. Any help would be great.

Thanks!

**Comment by johannesschobel** [comment]
Dear @alexmensak , not yet, sorry. Had a few other issues to deal with before i can work on the "deployment" problem for now.
But maybe @nikolasburk can give some more details on this issue? 

All the best,
Johannes

**Comment by MitchellMonaghan** [comment]
Is there any update on this? I'm having a lot of issues with a Quasar/Electron app. This is using webpack, typescript, electron-builder, and this has been a nightmare to get working. I still don't have functional code. As much as I love prisma I think I might have to not use it. Spent a few days trying to get this to work. Constantly blocked by "Could not open datamodel file "C:\\projects\\myApp\\dist\\electron\\Packaged\\win-unpacked\\resources\\app.asar\\schema.prisma" errors

**Comment by janpio** [comment]
Please open an issue about your problem then, we will be happy to look into that. Optimally include a small reproduction we can try out.

**Comment by nvitaterna** [comment]
> yeah, i am running into the same issue when using `nrwl/nx` as monorepository framework and `nestjs` for my API.
> 
> NX bundles the entire API into one large js file that is served. However, the `schema.prisma` file is not "copied" to the `dist/api` folder, hence, resulting in the previously mentioned error.
> 
> I am using the latest version of
> 
> * nrwl/nx
> * nestjs
> * prisma
> 
> Are there any suggestions? besides "manually" copying the files (or via npm `postXXX` scripts?)
> 
> All the best and thanks for your help,
> Johannes

I added this to the `assets` array in the `workspace.json` folder for my `api` app. Just change the input to the locatin of your prisma client.
```json
{
  "glob": "schema.prisma",
  "input": "libs/prisma-client/src",
  "output": ""
}
```
While this just copies the schema - it does it using the built-in builder so I feel like it's sufficient when it comes to nx.

**Comment by BetterLuke** [comment]
> ```json
> {
>   "glob": "schema.prisma",
>   "input": "libs/prisma-client/src",
>   "output": "."
> }
> ```

Thanks Bro. you save me a day~

**Comment by fivethreeo** [comment]
If using razzle, add a copyplugin

https://github.com/fivethreeo/razzle-prisma-graphqlcodegen-urql-formik-jwt/blob/main/razzle.config.js#L31

**Comment by reconbot** [comment]
I have a similar issue with esbuild, I'd like to specify the location for the different environments in my toolchain.

**Comment by vieirai** [comment]
Very annoyed to see that we still can't configure this.
As it stands, AWS SAM does not support bundling external files, as such the only way to load these is with a lambda layer, but those get mounted at `/opt` and not being able to specify an arbitrary location for `schema.prisma` is a deal breaker for anyone using AWS SAM.

Surely this shouldn't even be a difficult feature to implement, and all it takes is a quick browse and you'll find tons of posts asking about this exact problem.

**Comment by janpio** [comment]
Could you please open a new issue @vieirai and describe the full problem with AWS SAM? I don't think we know about that yet, and doing so would potentially allow us to understand that problem to suggest workarounds - or influence our opinion on this issue if this is really the only option. Thanks.

**Comment by vieirai** [comment]
@janpio done: https://github.com/prisma/prisma/issues/20199

**Comment by JustinMartinDev** [comment]
Hello, 

I was looking for a fix and after inspecting the generated prisma client code and I found a way to do it. I base the `config.dirname` on the `output` location.

In generated prisma client `index.js` update the code section about `alternativePath`


old :
```javascript
if (!fs.existsSync(path.join(__dirname, 'schema.prisma'))) {
  const alternativePaths = [
    "prisma/generated",
    "generated",
  ]; 
  
  const alternativePath = alternativePaths.find((altPath) => {
    return fs.existsSync(path.join(process.cwd(), altPath, 'schema.prisma'))
  }) ?? alternativePaths[0]

  config.dirname = path.join(process.cwd(), alternativePath)
  config.isBundled = true
}
``` 

new :
```javascript
if (!fs.existsSync(path.join(__dirname, 'schema.prisma'))) {
  const alternativePaths = [
    "prisma/generated",
    "generated",
  ]; 
   
  const alternativePath = alternativePaths.find((altPath) => {
    return fs.existsSync(path.join(process.cwd(), altPath, 'schema.prisma'))
  }); 

  if(alternativePath) {
    config.dirname = path.join(process.cwd(), alternativePath)
  } else if(config?.generator?.output?.value) {
    config.dirname = config.generator.output.value;
  } else {
    config.dirname = path.join(process.cwd(), alternativePaths[0]);
  }

  config.isBundled = true  
}
```
